### PR TITLE
Reworked routing algorithm

### PR DESCRIPTION
**Pull request by pm47**

### CODE REVIEW

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala [hunk 1]
```diff
@@ -529,37 +557,74 @@ object Router {
...
+    * @param withoutUpdates those will be removed before computing the route, and added back after so that g is left unchanged
+    * @return
+    */
+  def findRoute(g: DefaultDirectedWeightedGraph[PublicKey, DescEdge], localNodeId: PublicKey, targetNodeId: PublicKey, withUpdates: Map[ChannelDesc, ChannelUpdate] = Map.empty, withoutUpdates: Map[ChannelDesc, ChannelUpdate] = Map.empty): Try[Seq[Hop]] = Try {
+    if (localNodeId == targetNodeId) throw CannotRouteToSelf
     try {
+      withUpdates.foreach { case (d, u) => addEdge(g, d, u) }
+      withoutUpdates.foreach { case (d, _) => removeEdge(g, d) }
+
       val route_opt = Option(DijkstraShortestPath.findPathBetween(g, localNodeId, targetNodeId))
```
[the remark this answers is not in the document]
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]: problem is that the graph isn't bidirectional right?

[the remark this answers is not in the document]
Comment (btcontract) [reply eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]: It doesn't have to be. `BidirectionalDijkstraShortestPath` is a runtime optimization which runs two parallel searches on original graph and it's reversed copy so result should be the same.

[the remark this answers is not in the document]
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]: Ok I'll look into it 👍

[the remark this answers is not in the document]
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]: Apparently `BidirectionalDijkstraShortestPath` and `DirectedWeightedPseudograph` are incompatible :-(

**Pull request merged**